pkg/cmd/experimental/buildchain/buildchain.go

```diff
@@ -0,0 +1,339 @@
+
+    # Build and output the dependency tree in DOT syntax
+    $ openshift ex build-chain [image-repository]:[tag] -o dot
+
+    # Build dependency trees for all image repositories in the current namespace
+    $ openshift ex build-chain
+
+    # Build dependency trees for all image repositories across all namespaces
+    $ openshift ex build-chain --all
+`
+
+// ImageRepo is a representation of a node inside a tree
+type ImageRepo struct {
+	Name      string       `json:"name"`
+	Namespace string       `json:"namespace"`
+	Tags      []string     `json:"tags,omitempty"`
+	Edges     []*Edge      `json:"edges,omitempty"`
+	Children  []*ImageRepo `json:"children,omitempty"`
+}
+
+// String helps in dumping a tree in AST format
+func (root *ImageRepo) String() string {
+	tree := ""
+	tree += root.Name
+
+	for _, n := range root.Children {
+		tree += "(" + n.String() + ")"
+	}
+
+	return tree
+}
+
+// Edge represents a build configuration relationship
+// between two nodes
+//
+// Note that this type has no relation with the dot.Edge
+// type
+type Edge struct {
+	Name string `json:"buildConfig"`
+	To   string `json:"to"`
+}
+
+// NewEdge adds a new edge on a parent node
+//
+// Note that this function has no relation
+// with the dot.Edge type
+func NewEdge(name, to string) *Edge {
+	return &Edge{
+		Name: name,
+		To:   to,
+	}
+}
+
+// invalidRepoTagErr is returned when an invalid image repository and tag
+// combination has been passed by the user
+var invalidRepoTagErr = fmt.Errorf("invalid [imageRepository]:[tag] input")
+
+// NewCmdBuildChain implements all the necessary functionality for the build-chain module
+func NewCmdBuildChain(f *clientcmd.Factory, parentName, name string) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   fmt.Sprintf("%s [image-repository] [OPTIONS]", name),
+		Short: "Output build dependencies of a specific image repository",
+		Long:  longDescription,
+		Run: func(cmd *cobra.Command, args []string) {
+			all := cmdutil.GetFlagBool(cmd, "all")
+			if len(args) > 1 {
+				usageError(cmd, "Must pass nothing, an image repository name, or specify the --all flag")
+			}
+
+			osc, kc, err := f.Clients(cmd)
+			checkErr(err)
+
+			defaultNamespace, err := f.DefaultNamespace(cmd)
+			checkErr(err)
+
+			namespaces := make([]string, 0)
+			if all {
+				nsList, err := kc.Namespaces().List(labels.Everything())
+				checkErr(err)
+				for _, ns := range nsList.Items {
+					namespaces = append(namespaces, ns.Name)
+				}
+			} else {
+				namespaces = append(namespaces, defaultNamespace)
+			}
+
+			// Get all build configurations
+			buildConfigList := make([]buildapi.BuildConfig, 0)
+			for _, namespace := range namespaces {
+				cfgList, err := osc.BuildConfigs(namespace).List(labels.Everything(), labels.Everything())
+				checkErr(err)
+				buildConfigList = append(buildConfigList, cfgList.Items...)
+			}
+
+			// Get all image repositories from build configurations
+			repos := getRepos(buildConfigList, defaultNamespace)
```

Comment:
I am using all the repos in order to validate that the specified image repository is valid (ie. part of the buildConfigs)